File: repeatedDNAsequences.py
#O(n) time and O(n) space solution
class Solution:
    def findRepeatedDnaSequences(self, s: str) -> List[str]:
        sequence_seen = set()
        result = set()
        for i in range(len(s) - 9):
            currentSequence = s[i : i + 10]
            if currentSequence in sequence_seen:
                result.add(currentSequence)
            else:
                sequence_seen.add(currentSequence)
        return result

File: RepeatedDNASequences.cpp
// Time Complexity -> O(10*n) ~ O(n)
// Space Complexity -> O(10*n) ~ O(n)

class Solution {
public:
    vector<string> findRepeatedDnaSequences(string s) {
        int n = s.length();
        set<string> allSubStrings;
        set<string> res;
        
        for(int i = 0; i < n; i++){
            string curr = s.substr(i, 10);
            if(allSubStrings.count(curr))
                res.insert(curr);
            allSubStrings.insert(curr);
        }
        
        vector<string> repeats;
        for(auto itr = res.begin(); itr != res.end(); ++itr)
            repeats.push_back(*itr);
        
        return repeats;
    }
};